src/stdlib/Conversions.scillib

```diff
@@ -0,0 +1,129 @@
+(* ******************************************************************************** *)
+(*   This file is part of scilla.                                                   *)
+(*                                                                                  *)
+(*   Copyright (c) 2018 - present Zilliqa Research Pvt. Ltd.                        *)
+(*                                                                                  *)
+(*   scilla is free software: you can redistribute it and/or modify it under the    *)
+(*   terms of the GNU General Public License as published by the Free Software      *)
+(*   Foundation, either version 3 of the License, or (at your option) any later     *)
+(*   version.                                                                       *)
+(*                                                                                  *)
+(*   scilla is distributed in the hope that it will be useful, but WITHOUT ANY      *)
+(*   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR  *)
+(*   A PARTICULAR PURPOSE.  See the GNU General Public License for more details.    *)
+(*                                                                                  *)
+(*   You should have received a copy of the GNU General Public License along with   *)
+(*   scilla.  If not, see <http://www.gnu.org/licenses/>.                           *)
+(* ******************************************************************************** *)
+scilla_version 0
+
+import IntUtils
+
+library Conversions
+
+(* Is the ByStr long enough for substr, starting at pos and len. *)
+let is_long_enough : ByStr -> Uint32 -> Uint32 -> Bool =
```

Comment:
Hmm, I would say that an idiomatic way in Scilla would be to return an `Option` type (`Some bstr` would represent the substring) instead of a Boolean.

Reply:
That is what `extract_scillaval` is doing. This is just an accessory for `extract_scillaval`, to do the numerical calculation. I could move it inside `extract_scillaval` if there's no other use for it.

Reply:
Yes, but then you run `builtin substr` after that check. And this will be the usual case.
I'd propose to have a "safe" function which combines both the check and the call to `substr` builtin, like this one:
```
(* does not throw exceptions *)
let osubstr : ByStr -> Uint32 -> Uint32 -> Option ByStr =
  fun (bs : ByStr) =>
  fun (pos : Uint32) =>
  fun (len : Uint32) =>
    let length = builtin strlen bs in
    let shorter = uint32_le len length in
    match shorter with
    | True =>
      let delta = builtin sub length len in
      let safe_pos = uint32_le pos delta in
      match safe_pos with
      | True => let substr = builtin substr bs pos len in
                Some {ByStr} substr
      | False => None {ByStr}
      end
    | False => None {ByStr}
    end
```
Notice that I have two checks instead of one: that's because I think the following test should work and not fail with integer overflow:
```
let bs = 0x00 in
let pos = Uint32 4294967295 in (* max Uint32 *)
let len = Uint32 1 in
osubstr bs pos len
```

Reply:
Why is this a concern at all? Likely, `is_long_enough` is just going to go inside `extract_scillaval`.

Reply:
Then it's fine as long as it's not part of the interface. Although, we could have provided safer versions of the builtins, since it looks like we extend stdlib anyways.

Reply:
Okay, I just pushed a commit that includes your `osubstr` and uses it ! but it won't be in `master` for a while :)

Reply:
Awesome! :)